test/reshape.jl

```diff
@@ -507,6 +489,97 @@ end
+              DataFrame(id=1:3, a=1:3:7, b=2:3:8, c=3:3:9)
+        @test unstack(df2, [:id, :id2], :var, :val) == unstack(df2, :var, :val) ==
+              DataFrame(id=1:3, id2=1:3, a=1:3:7, b=2:3:8, c=3:3:9)
+        levels!(df2.id, [10, 2, 11, 3, 1, 12])
+        levels!(df2.var, ['x', 'b', 'y', 'c', 'a', 'z'])
+        @test unstack(df2, :id, :var, :val) ==
+              DataFrame(id=1:3, b=2:3:8, c=3:3:9, a=1:3:7)[[2,3,1], :]
+        @test unstack(df2, [:id, :id2], :var, :val) == unstack(df2, :var, :val) ==
+              DataFrame(id=1:3, id2=1:3, b=2:3:8, c=3:3:9, a=1:3:7)[[2,3,1], :]
+    end
+
+    df = DataFrame(id=repeat(1:3, inner=3),
+                   a=repeat(1:3, inner=3),
+                   var=repeat('a':'c', 3),
+                   val=1:9)
+    @test unstack(df, :id, :var, :val) == DataFrame(id=1:3, a=1:3:7, b=2:3:8, c=3:3:9)
+    @test_throws ArgumentError unstack(df, :a, :var, :val)
+    @test_throws ArgumentError unstack(df, [:id, :a], :var, :val)
+
+    df = DataFrame(id=repeat(1:3, inner=3),
+                   id2=repeat(1:3, inner=3),
+                   var=repeat('a':'c', 3),
+                   val=1:9)
+    df[4, 1:2] .= 1
+    @test_throws ArgumentError unstack(df, :id, :var, :val)
+    @test_throws ArgumentError unstack(df, [:id, :id2], :var, :val)
+    @test unstack(df, :id, :var, :val, allowduplicates=true) ≅
+          DataFrame(id=1:3, a=[4, missing, 7], b=2:3:8, c=3:3:9)
+    @test unstack(df, [:id, :id2], :var, :val, allowduplicates=true) ≅
+          DataFrame(id=1:3, id2=1:3, a=[4, missing, 7], b=2:3:8, c=3:3:9)
+
+    df = DataFrame(id=repeat(1:3, inner=3),
+                   id2=repeat(1:3, inner=3),
+                   var=repeat('a':'c', 3),
+                   val=1:9)
+    allowmissing!(df, :var)
+    df.var[4] = missing
+    @test_throws ArgumentError unstack(df, :id, :var, :val)
+    @test_throws ArgumentError unstack(df, [:id, :id2], :var, :val)
+    @test unstack(df, :id, :var, :val, allowmissing=true) ≅
+          DataFrame(id=1:3, a=[1, missing, 7], b=2:3:8, c=3:3:9, missing=[missing, 4, missing])
+    @test unstack(df, [:id, :id2], :var, :val, allowmissing=true) ≅
+          DataFrame(id=1:3, id2=1:3, a=[1, missing, 7], b=2:3:8, c=3:3:9, missing=[missing, 4, missing])
+end
+
+# test scenario when sorting fails both in grouping and in variable
+struct A_TYPE
+    x
+end
+
+@testset "additional unstack tests not sortable" begin
+    df = DataFrame(id=repeat(A_TYPE.([2, 1, 3]), inner=3),
+                   id2=repeat(A_TYPE.([2, 1, 3]), inner=3),
+                   var=repeat(A_TYPE.([3, 2, 1]), 3),
+                   val=1:9)
+    @test unstack(df, :id, :var, :val, renamecols=x -> Symbol(:x, x.x)) ==
+          DataFrame(id=A_TYPE.([2, 1, 3]), x3=1:3:7, x2=2:3:8, x1=3:3:9)
+    @test unstack(df, [:id, :id2], :var, :val, renamecols=x -> Symbol(:x, x.x)) ==
+          DataFrame(id=A_TYPE.([2, 1, 3]), id2=A_TYPE.([2, 1, 3]), x3=1:3:7, x2=2:3:8, x1=3:3:9)
+end
```

Comment:
We don't need this if we don't sort, right?

Reply:
we test in this case we order things in the order of first appearance.